apps/src/javalab/Playground.js

```diff
@@ -128,38 +128,59 @@ export default class Playground {
   }
 
   addTextItem(itemData) {
-    if (this.isGameOver) {
-      // can't add new items if the game is over
+    if (this.isGameOver || this.itemExists(itemData)) {
```

Comment:
I'm noticing a lot of checks for correct state on the frontend, but the backend is the source of truth for Playground. How are we managing error handling between the two?

Part of this question is coming from what I'm hoping is the correct understanding about how Neighborhood works. There, we'd allow most things to happen on the front end and it was the back end that was responsible for maintaining boundaries, obstacles, etc. This allowed us to provide descriptive error messages and trigger exceptions from the student's code. It also made things much more obvious when we wrote a bug - there was a point when the painter was regularly going off the screen.

Reply:
I moved this conversation to slack for better tracking: https://codedotorg.slack.com/archives/C01EF4GJ9GE/p1633020656468500